Draft: gatsby wpgraphql theme and source

TylerBarnes opened the pull request:
Since fri I built out the initial code for `gatsby-source-wpgraphql` and laid the framework for incremental builds and preview in WP.
The source plugin generates gql queries from introspection, downloads everything from WP (as long as it's in a list of nodes), and makes nodes in Gatsby from the response.
On subsequent builds, Gatsby asks WP what's changed by sending a gql request to query a log of recent actions on the WP site that I set up in our [WPGatsby](https://github.com/jasonbahl/wp-gatsby/) plugin. 
The timestamp that the last build was completed is sent as an argument in that query and only updates since that time are returned. 
The source plugin then loops through those updates and either deletes, creates, or updates nodes.
Once I built that I realized it would be really good for incremental builds and preview so I set up an interval to refetches this new field type over and over and apply those updates during `develop`. Since the query will usually return an empty array, we can do this repeatedly. There's still a ton of work to be done and some of this code isn't very efficient but this seems like a good start!

freiksenet commented on examples/using-gatsby-source-wpgraphql/README.md:
```diff
@@ -0,0 +1,11 @@
+# gatsby-source-graphql example with GraphCMS
```
[nit] old readme

freiksenet commented on packages/gatsby-source-wpgraphql/CHANGELOG.md:
```diff
@@ -0,0 +1,260 @@
+# Change Log
```
[nit] Old changelog

freiksenet commented on packages/gatsby-source-wpgraphql/README.md:
```diff
@@ -0,0 +1,159 @@
+# gatsby-source-graphql
```
[nit] old readme

KyleAMathews commented on packages/gatsby-source-wpgraphql/src/gatsby-node/source-nodes/wp-actions/delete.js:
```diff
@@ -0,0 +1,24 @@
+  const node = await getNode(nodeId)
+
+  if (node) {
+    await actions.touchNode({ nodeId })
```
curious why you're touching then deleting the node?

muescha commented on packages/gatsby-source-wpgraphql/src/gatsby-node/create-pages.js:
```diff
@@ -0,0 +1,34 @@
+        component: template,
+        path: node.path,
+        context: {
+          ID: node.id,
```
others have here a lowercased...

```suggestion
          id: node.id,
```

muescha commented on examples/using-gatsby-source-wpgraphql/src/templates/index.js:
```diff
@@ -0,0 +1,49 @@
+}
+
+export const query = graphql`
+  query index($ID: String!) {
```
```suggestion
  query index($id: String!) {
```

muescha commented on examples/using-gatsby-source-wpgraphql/src/templates/index.js:
```diff
@@ -0,0 +1,49 @@
+
+export const query = graphql`
+  query index($ID: String!) {
+    wpContent(id: { eq: $ID }) {
```
```suggestion
    wpContent(id: { eq: $id }) {
```

muescha commented on packages/gatsby-source-wpgraphql/gatsby-node/create-pages.js:
```diff
@@ -0,0 +1,37 @@
+      component: template,
+      path: node.path,
+      context: {
+        ID: node.id
```
```suggestion
        id: node.id
```

muescha commented:
where did the query goes?
i can not find any code snippets about this...

```graphql
    query GET_ACTION_MONITOR_ACTIONS($since: Float!) {
      actionMonitorActions(where: {sinceTimestamp: $since}) {
        nodes {
          referencedPostID
          referencedPostStatus
          referencedPostGlobalRelayID
          referencedPostSingleName
          referencedPostPluralName
          actionType
        }
      }
    }
```

TylerBarnes commented:
> where did this graphql comes from?
> i can not find any code snippets about this...
> 
> ```graphql
>     query GET_ACTION_MONITOR_ACTIONS($since: Float!) {
>       actionMonitorActions(where: {sinceTimestamp: $since}) {
>         nodes {
>           referencedPostID
>           referencedPostStatus
>           referencedPostGlobalRelayID
>           referencedPostSingleName
>           referencedPostPluralName
>           actionType
>         }
>       }
>     }
> ```
> 
> PS: is [jasonbahl/wp-gatsby](https://github.com/jasonbahl/wp-gatsby/) a private repo?

@muescha it is in a private repo! I forgot about that when I opened this.
The `actionMonitorActions` field is defined in WPGatsby and returns the latest changes since the last build time or the last time the plugin checked for changes. Right now it's set up to do long polling as that query returns an empty array 99% of the time and will only return a very small amount of data when a change has occurred in WP. We're planning on looking at adding subscriptions instead of long polling but it will likely depend on an external service since most PHP servers don't support websockets. Long polling will always be supported as a works-anywhere option.

Also, a few of us had a conversation about this PR and some changes will be:
- Move WPGatsby into the Gatsby repo and into this PR
- this will be `gatsby-source-wordpress@v4` instead of `gatbsy-source-wpgraphql`
- There will be a Github project to track the work done in this PR (or branch) and everything will be broken into subtasks for anyone that's interested in helping out with the direction or writing some code

To add to the above - this PR is very much a WIP. I threw it together quickly to outline in working code how I think things should function based on discussions a few of us on the WP Gatsby team have had over the last few weeks & based on my experimentation with Gatsby/WP that I've been doing over the last year and a half.
Much of the code needs to be expanded on or refactored to be production ready.

muescha commented:
Personal I would not like the `PostID` better `PostId`

Also.normal.in other WordPress plugins there are only short names


So I expect more this

```graphql

related {
   post {
      id
     }
  }
```

muescha commented:
also there are some inconsistent naming `singleNames` and `singularNamr` and also assign `singleName` to a name without marking it with single/singular

Sometimes n

TylerBarnes commented:
> Personal I would not like the `PostID` better `PostId`
> 
> Also.normal.in other WordPress plugins there are only short names
> 
> So I expect more this
> 
> ```graphql
> related {
>    post {
>       id
>      }
>   }
> ```

The schema is based on the WPGraphQL schema, in the interest of keeping maintenance simple, we should modify the schema on the Gatsby side the least amount possible to not make the code overly complex. There will already need to be a lot done to modify the WPGQL schema, so I would leave small things like this out.
Additionally we're working on a `v1.0.0` release for WPGraphQL which will modify the schema on the WP side and not everything that currently exists will remain the same. 
V1 of WPGraphQL will be released alongside this plugin and there will be some back and forth between the two plugins as the two each progress towards a V1, but for the general field shape and field names like ID fields, I don't think that will be changing as many existing production apps and sites use WPGraphQL.

TylerBarnes commented:
I'll be gone for the next week. FYI if anyone pulls this down while I'm gone out of curiosity, the example is actually broken from 78c7485 onwards since it's using types from WPGQL now instead of inference, so the `WpContent` type which was being created via inference and used in the example is now missing.

michaellopez commented:
We're giving this a test run with 213000+ posts spread across 70+ blogs.

What we noticed immediately was that there already exists a [`WPGatsby`](https://wordpress.org/plugins/wp-gatsby/) plugin, so there is a naming collision. Making it look like there is an update:

![image](https://user-images.githubusercontent.com/487039/69513731-09394980-0f49-11ea-8169-7624e1bb1a91.png)

Maybe you need to try and get ownership of that plugin name/namespace or rename the plugin?

TylerBarnes commented:
@michaellopez that's awesome! We talked with the author of that plugin and he was happy to pass it off to us. We probably wont be publishing WPGatsby to the WP repo for a while so that notice will be there for a bit.
FYI things are currently broken in this PR as I gradually implement more GQL functionality, so the code is far from being in a good place still. I did test it out on a site with 25k pages (when the demo was still functional) and it worked like a charm! More work is going to be put into ensuring the plugin will run on all types of servers as well.

michaellopez commented:
@TylerBarnes Thanks for the info. Great that you are proactive in the matter.

As I mentioned in #19292 we are in the middle of a project with a deadline on new years. We are trying to figure out the best way to source content from WordPress into Gatsby. Taking inspiration (a.k.a. stealing proudly) from this draft PR we will look into making a limited approach for our use case. Our use case mostly only needs to deal with posts and authors in a simple manner so hopefully our work will give us a better understanding of the inner workings in this PR so that we can help/contribute back when this PR gets to a state where contributions can more easily be made.

If there is anything you think we can help out with to move this PR along let me know 💪 

I dare to say that a lot of the internets are rooting for your success 😍

TylerBarnes commented:
Thanks @michaellopez ! 😄 I'll keep you in the loop!!
